### ISSUE TITLE
ember new app won't start UnhandledPromiseRejectionWarning

### ISSUE THREAD

**Issue by eccegordo** [ember-cli/ember-cli#9074]
Hi

This is probably related to some sub dependency. But seems to affect ember CLI in the most basic use case. For some reason I cannot get a brand new app to start or compile. I have tried different versions of node and I keep running into an issue with

UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'graph' of undefined

---

Output from `ember version --verbose && npm --version && yarn --version`:
```
ember-cli: 3.16.0
http_parser: 2.9.3
node: 10.19.0
v8: 6.8.275.32-node.55
uv: 1.28.0
zlib: 1.2.11
brotli: 1.0.7
ares: 1.15.0
modules: 64
nghttp2: 1.39.2
napi: 5
openssl: 1.1.1d
icu: 64.2
unicode: 12.1
cldr: 35.1
tz: 2019c
os: darwin x64
6.13.4
zsh: command not found: yarn
```

Here is a full trace of the steps I follow.

```
~/LocalOnly.nosync/tempcode
❯ ember -v
ember-cli: 3.16.0
node: 10.19.0
os: darwin x64

~/LocalOnly.nosync/tempcode
❯ ember new app1
installing app
Ember CLI v3.16.0

✨  Creating a new Ember app in /Users/gordon/LocalOnly.nosync/tempcode/app1:
  create .editorconfig
  create .ember-cli
  create .eslintignore
  create .eslintrc.js
  create .template-lintrc.js
  create .travis.yml
  create .watchmanconfig
  create README.md
  create app/app.js
  create app/components/.gitkeep
  create app/controllers/.gitkeep
  create app/helpers/.gitkeep
  create app/index.html
  create app/models/.gitkeep
  create app/router.js
  create app/routes/.gitkeep
  create app/styles/app.css
  create app/templates/application.hbs
  create config/environment.js
  create config/optional-features.json
  create config/targets.js
  create ember-cli-build.js
  create .gitignore
  create package.json
  create public/robots.txt
  create testem.js
  create tests/helpers/.gitkeep
  create tests/index.html
  create tests/integration/.gitkeep
  create tests/test-helper.js
  create tests/unit/.gitkeep
  create vendor/.gitkeep

🚧  Installing packages... This might take a couple of minutes.
npm: Installed dependencies

🎥  Initializing git repository.
Git: successfully initialized.

🎉  Successfully created project app1.
👉   Get started by typing:

  $ cd app1
  $ npm start

Happy coding!

~/LocalOnly.nosync/tempcode 36s
❯ cd app1

~/LocalOnly.nosync/tempcode/app1 master
❯ ls
README.md		node_modules		testem.js
app			package-lock.json	tests
config			package.json		vendor
ember-cli-build.js	public

~/LocalOnly.nosync/tempcode/app1 master
❯ npm start

> app1@0.0.0 start /Users/gordon/LocalOnly.nosync/tempcode/app1
> ember serve


Build successful (8036ms) – Serving on http://localhost:4200/



Slowest Nodes (totalTime => 5% )              | Total (avg)
----------------------------------------------+---------------------
Babel: @ember/test-helpers (2)                | 1477ms (738 ms)
Package /assets/vendor.js (1)                 | 1092ms
BroccoliRollup (6)                            | 780ms (130 ms)
Babel: @ember-data/store (2)                  | 661ms (330 ms)
Babel: ember-inflector (1)                    | 479ms

(node:5176) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'graph' of undefined
    at promiseFinally.watcher.currentBuild.then.hash (/Users/gordon/LocalOnly.nosync/tempcode/app1/node_modules/ember-cli/lib/models/watcher.js:64:29)
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:5176) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:5176) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

```

**Comment by paulyoder** [comment]
I'm running into the exact same issue (my only difference is I'm using Yarn instead of NPM)

**Comment by stefanpenner** [comment]
This is likely due to the broccoli release `v3.2.0`, i caught some issues with it and have since release `v3.2.1` with those fixes, and `v3.2.2` with upgrades.

Can you run `yarn why broccoli` and share the exact version you are using?

**Comment by stefanpenner** [comment]
I've reproduced the problem locally, investigating.

**Comment by eccegordo** [comment]
@stefanpenner When you find the fix can you do a quick post mortem to help me understand the basic gist of issue? Mostly curious to better understand. Not even sure how to upgrade broccoli and/or navigate some of the other internal deps. I am approaching this from a very naive set of assumptions:

nvm install node version
npm install -g ember-cli
ember new app
boom!

**Comment by mynoheart** [comment]
Temporary fix is ```npm install broccoli@3.2.0``` in new project.

**Comment by mynoheart** [comment]
This helped for me.

**Comment by stefanpenner** [comment]
@mynoheart thanks, i was just about to post that as a work around.

**Comment by stefanpenner** [comment]
@eccegordo ya, i'll do that.

Just made myself some coffee, debugging this more seriously now.

**Comment by stefanpenner** [comment]
The TL:DR broccoli itself hasn't been release in some time, I was fixing an ember-cli bug in broccoli. Spot checking it, with the known issue. Once I released broccoli I noticed several regression had occurred.

https://github.com/broccolijs/broccoli/pull/453 + https://github.com/broccolijs/broccoli/pull/454 

...

^^ the issue of this thread is surprising, if it is not the last one I'll revert the release of broccoli.

**Comment by stefanpenner** [comment]
Found the problem, I believe I have it fixed locally, writing tests now to confirm and prepare the broccoli PR....

**Comment by stefanpenner** [comment]
fix PR: https://github.com/broccolijs/broccoli/pull/457

**Comment by stefanpenner** [comment]
broccoli 3.3.3 has been released, addressing this issue.

Just ran the above reproduction to confirm the issue as reported has been addressed.

---

Short TL;DR / Post Moterm.

There is a piece of state passed around between the broccoli watcher, the builder, ember-cli, back to the watcher, and then back to ember-cli. This piece of state is called the buildAnnotation. The buildAnnotation holds descriptive details about the current build, such as:

```
* type: 'initial' | 'rebuild'
* reason: 'watcher' | 'build'
* primaryFile?:  filePath
* changedFiles: filePaths[]
* filePath?: filePath
* graph?: instrumentation metadata provided by ember-cli
```

Needless to say, this little object has grown in importance since initially implemented, but broccoli's tests did not sufficiently test the rather complicated lifecycle and transportation of this object.

So today, when I refactored old promise code to modern async/wait to provide better stack traces for ember-cli users. I inadvertently awaited the build, when I should have returned it. This caused `watcher.currentBuild` to initially fulfill with undefined, rather then the build annotation. Doing so resulted in ember-cli attempting to read data, the Heimdall graph, off undefined.

To address this, I have backfilled the necessary tests within broccoli.
